### PR TITLE
Add Navigation and documentation

### PR DESCRIPTION
**Pull request by rogeriochaves**
Hello,

I really want to use this, so I'm trying to make the necessaries changes to make it usable at least with elm-github-install. I've added documentation, which I believe we can improve later with code examples, and renamed `_user$project` to `_xavh$elm-testable`.

The reason to use `xavh` and not `avh` is because the native modules are loaded in alphabetical order, so "avh" was causing everything to fail.

It would be much better if we could not depend on loading order, with a solution like this: https://github.com/elm-community/webgl/pull/14/files

But idk if this will work for elm-testable, what do you think? Any ideas?

### CODE REVIEW

**Comment by rogeriochaves** [comment]
I've also added the code needed for working with elm-lang/navigation, I believe this lib is now ready to be used, the code might be not in its best fit yet, but it should be enough for people to test it and report bugs

FYI i'm already using it with elm-github-install pointing to my branch: https://github.com/CaronaBoard/caronaboard/tree/master/tests/Integration

It's not a very big project, but not very small either, and elm-testable is working perfectly!

**Comment by rogeriochaves** [comment]
I believe travis build needs to be reran without cache

**Comment by rogeriochaves** [comment]
@avh4 ping

**Comment by rogeriochaves** [comment]
I've also updated to the latest elm html test version

**Comment by rogeriochaves** [comment]
Changed xavh4/elm-testable back to user/project

**Comment by avh4** [comment]
Cleared the Travis cache.  (did you understand why the elm-stuff needed to be cleared?)

File: examples/NavigationSample.elm [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+type Msg
+    = UrlChange Navigation.Location
+    | Go String
+
+
+
+{- We are just storing the location in our history in this example, but
+   normally, you would use a package like evancz/url-parser to parse the path
+   or hash into nicely structured Elm values.
```
Comment (avh4) [inline examples/NavigationSample.elm]: Doesn't necessarily need to be done for this PR, but I think it would be good to make this example a bit more realistic and to the URL parsing.

File: examples/tests/NavigationSampleTests.elm [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        [ test "renders navigation history" <|
+            \() ->
+                NavigationSample.program
+                    |> start
+                    |> simulate (find [ class "bears" ]) Event.click
+                    |> simulate (find [ class "cats" ]) Event.click
+                    |> expectView
+                    |> find [ class "history" ]
+                    |> Expect.all
+                        [ has [ text "#bears" ]
```
Comment (avh4) [inline examples/tests/NavigationSampleTests.elm]: when this is run in real life, won't the text be the entire page URL, not simply the hash part?

File: examples/tests/NavigationSampleTests.elm [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                        , has [ text "#cats" ]
+                        ]
+        , test "works for user initiated navigation" <|
+            \() ->
+                NavigationSample.program
+                    |> start
+                    |> navigate "#cats"
```
Comment (avh4) [inline examples/tests/NavigationSampleTests.elm]: Same question as above; I don't think we can limit the API to only working with the hash part of URLs.  Does the API work with full URLs?  If so, I think the example should show what that would look like.  If not, we should think about how to deal with that.

File: src/Native/TestContext.js [hunk 4]
```diff
@@ -121,7 +121,8 @@ var _user$project$Native_TestContext = (function () { // eslint-disable-line no-
         init: app.init,
         update: program.elmTestable.update,
         subscriptions: program.elmTestable.subscriptions,
-        view: program.elmTestable.view
+        view: program.elmTestable.view,
+        locationToMessage: program.elmTestable.locationToMessage || { ctor: 'Nothing' }
```
Comment (avh4) [inline src/Native/TestContext.js]: Do you think it's best to put the fallback Nothing here, or should we instead update the Html.program intercept to add it there?

**Review by avh4 — COMMENTED** [review]
(Didn't finish reviewing the impl yet, but posting a couple comments to start with.)

File: src/Native/Testable/Navigation.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        _0: {
+          ctor: ctor,
+          _0: data
+        },
+        _1: function (data) { return { ctor: 'Success', _0: data } }
+      }
+  }
+}
+
+_elm_lang$navigation$Navigation$pushState = setItUp(
```
Comment (avh4) [inline src/Native/Testable/Navigation.js:39]: Does it not work if we intercept `Native_Navigation$pushState` instead of `Navigation$pushState`?

File: src/Testable/Navigation.elm [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+
+init : History
+init =
+    ( 0, [ initialLocation ] )
+
+
+initialLocation : Location
+initialLocation =
+    getLocation "https://elm.testable/"
```
Comment (avh4) [inline src/Testable/Navigation.elm]: Is there a nice way we could make the API require the test author to provide the initial location iff the program being tested is a `Navigation.program` and not have a default provided?

File: src/Testable/Navigation.elm [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+import List.Extra
+import
+    -- This "unused" import is required for the native bindings
+    Native.Testable.Navigation
+import Navigation exposing (Location)
+import Regex exposing (HowMany(..), find, regex)
+
+
+type alias History =
+    ( Int, List Location )
```
Comment (avh4) [inline src/Testable/Navigation.elm:12]: Ah, I was confused at first what the `Int` was for, but it makes sense from looking at the code. :)

1) I think the data structure here should be made private instead of being a type alias.

2) I think it would be ideal to use a list zipper instead of (Int, List).  There is wernerdegroot/listzipper, but it doesn't have nice functions for adding to the list.  Maybe we should have a ListZipper module in this project for now that either wraps wernerdegroot/listzipper or implements just what we need.   (In any case, doing that refactoring doesn't need to be done in this PR.)

File: src/Native/Testable/Navigation.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+_elm_lang$navigation$Navigation$go = setItUp(
+  _elm_lang$navigation$Navigation$go,
+  mapNavigation('Jump')
+)
+
+_elm_lang$navigation$Navigation$setLocation = setItUp(
+  _elm_lang$navigation$Navigation$setLocation,
+  mapNavigation('Visit')
```
Comment (avh4) [inline src/Native/Testable/Navigation.js:56]: I think it would be preferable to have these ctors named after the functions they're intercepting, so `PushState/ReplaceState/Go/SetLocation` instead of `New/Modify/Jump/Visit`

File: src/Testable/Navigation.elm [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    , pathname =
+        if String.isEmpty pathname then
+            "/"
+        else
+            pathname
+    , search = matchOrEmptyAt 5
+    , hash = matchOrEmptyAt 6
+    , username = ""
+    , password = ""
+    }
```
Comment (avh4) [inline src/Testable/Navigation.elm:140]: Is there not another package that does this correctly?  If not, we should look at making/publishing that package (not necessary for this PR, though).

File: src/Testable/Navigation.elm [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+
+setLocation : String -> Navigation.Location -> Navigation.Location
```
Comment (avh4) [inline src/Testable/Navigation.elm]: The relevant spec for this appears to be https://www.w3.org/TR/WD-html40-970708/htmlweb.html

I think we should at least add a comment mentioning that and a TODO to someday review the implementation here against that spec.  (Not necessary for this PR, though, but I do think we should add that comment at least.)

File: src/Testable/Navigation.elm [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+setLocation url currentLocation =
+    let
+        nextHref =
+            if Regex.contains (regex "^/") url then
+                currentLocation.origin ++ url
+            else if Regex.contains (regex "^\\?") url then
```
Comment (avh4) [inline src/Testable/Navigation.elm:151]: This case (initial `?`) doesn't seem to be a legitimate use of URLs in HTML.  Does this actually work in practice?

File: tests/NavigationTests.elm [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+initWithStringFlags : String -> Navigation.Location -> ( Model, Cmd Msg )
+initWithStringFlags flags location =
+    init location
+
+
+programUpdate : Msg -> Model -> ( Model, Cmd Msg )
+programUpdate msg model =
+    case msg of
+        UrlChange location ->
+            ( { model | location = location, msgs = msg :: model.msgs }, Navigation.modifyUrl location.pathname )
```
Comment (avh4) [inline tests/NavigationTests.elm]: Might be clearer to call the `msgs` field `urlChanges` instead since it doesn't record all messages.

File: tests/NavigationTests.elm [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+
+init : Navigation.Location -> ( Model, Cmd Msg )
+init location =
+    ( { location = location, msgs = [] }, Cmd.none )
+
+
+initWithStringFlags : String -> Navigation.Location -> ( Model, Cmd Msg )
+initWithStringFlags flags location =
+    init location
```
Comment (avh4) [inline tests/NavigationTests.elm]: I think we should add a test case where the flags are actually used (probably just stored in the model and then verified)

File: tests/NavigationTests.elm [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        |> TestContext.startWithFlags "foo"
+
+
+all : Test
+all =
+    describe "Navigation TestContext"
+        [ test "starting a navigation program" <|
+            \() ->
+                sampleProgram
+                    |> expectHref "https://elm.testable/"
```
Comment (avh4) [inline tests/NavigationTests.elm]: Yeah, I commented above about this, but from seeing the tests, I definitely think we should add `TestContext.startWithLocation` and `TestContext.startWithFlagsAndLocation` instead of hardcoding a default URL, if that's possible.  (Not sure if that should take a `Location` value or a `String` URL that gets turned into the initial Location?  Probably the String is better?)

File: tests/NavigationTests.elm [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+            , test "erases forward history" <|
+                \() ->
+                    sampleProgram
+                        |> navigate "/foo"
+                        |> update (Back 1)
+                        |> navigate "/baz"
+                        |> update (Back 100)
+                        |> update (Forward 1)
+                        |> expectHref "https://elm.testable/baz"
+            ]
```
Comment (avh4) [inline tests/NavigationTests.elm]: I think we need to add a `Test.Navigation` API to simulate the user clicking back/forward in the web browser in addition to handling the Cmds in `Navigation`

Comment (avh4) [inline tests/NavigationTests.elm]: What happens in the real world if you go back or forward when you're at the beginning/end of the history?  (I'm assuming your implementation probably already works correctly for those cases), but I think we should probably have test cases demonstrating those scenarios.

File: tests/NavigationTests.elm [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+                        |> expectHref "https://elm.testable/foo"
+            , test "erases forward history" <|
+                \() ->
+                    sampleProgram
+                        |> navigate "/foo"
+                        |> update (Back 1)
+                        |> navigate "/baz"
+                        |> update (Back 100)
+                        |> update (Forward 1)
+                        |> expectHref "https://elm.testable/baz"
```
Comment (avh4) [inline tests/NavigationTests.elm]: What happens in the real world if you go back or forward when you're at the beginning/end of the history?  (I'm assuming your implementation probably already works correctly for those cases), but I think we should probably have test cases demonstrating those scenarios.

File: tests/NavigationTests.elm [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+            \() ->
+                sampleProgram
+                    |> update (Load "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "refresh does nothing" <|
+            \() ->
+                sampleProgram
+                    |> update Reload
+                    |> done
+        , describe "navigation simulation"
```
Comment (avh4) [inline tests/NavigationTests.elm]: I think we also need tests exercising the different kinds of URLs that are accepted (relative / fragment / absolute).  I think it's fine to do all those just with one of the functions, though (maybe pushState).

Comment (avh4) [reply tests/NavigationTests.elm]: Oh, nevermind! I see you have all those tests below in another file!  💯

File: tests/Testable/NavigationTests.elm [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+                            |> Expect.equal
+                                ( ( 1
+                                  , [ initialLocation
+                                    , getLocation "https://elm.testable/foo"
+                                    ]
+                                  )
+                                , TriggerLocationMsg (getLocation "https://elm.testable/foo")
+                                )
+                ]
+            ]
```
Comment (avh4) [inline tests/Testable/NavigationTests.elm:250]: I'm assuming all these tests were useful for while implementing.  I'm curious if you think they're still valuable to have around now that there are also tests in `NavigationTests` that I expect cover most if not all of the code tested here?  (I'm fine either way, but I'm curious what you think.)

**Comment by avh4** [comment]
Also, thanks for adding all the docs!

**Comment by avh4** [comment]
I finished reviewing all that code.  Thanks for all the work!  I left a few ideas and questions.  (If you don't find time to get back to this, I'll eventually get around to addressing them myself.)

File: examples/tests/NavigationSampleTests.elm (hunk 2 above)
Comment (rogeriochaves) [reply examples/tests/NavigationSampleTests.elm]: yes, but it does, it is actually not the whole url in this case, just the pathname + the hash, like this:

image

but, the `has [ text` matcher does not search for the exact whole string, it just tries to find this string piece anywhere. So for example, if I change this for some text that is not on the DOM, I can visualize the error and that's what I get:

image

You see, it is actually `/#cats` because the default `pathname` for the tests is `/`

I'll change this test to be more specific and include the `/`

File: examples/tests/NavigationSampleTests.elm (hunk 3 above)
Comment (rogeriochaves) [reply examples/tests/NavigationSampleTests.elm]: See answer above, yes, it should correctly work with full urls, I've tried to catch all the edge cases by testing it against a real project, and adding more tests

File: src/Testable/Navigation.elm (hunk 7 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm:12]: I've never actually used zippers before, so I'll leave this part for you ;)

File: src/Testable/Navigation.elm (hunk 9 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm:140]: Yeah, I though the same, this should be extracted to a separate package. I searched and the closest thing I found was http://package.elm-lang.org/packages/sporto/erl/latest, but it parses to its own type, not a elm-lang/navigation one.

Nevertheless I like to wait a little and see that this code really works for us before extracting to a package, it's quicker to change for now

File: src/Testable/Navigation.elm (hunk 10 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm]: agreed, I missed checking this agains a spec, but I also got lazy. Using regexes does open spaces for a lot of edge cases and I don't know if I caught them all (surely not), I'll add the TODO

File: src/Testable/Navigation.elm (hunk 11 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm:151]: yes, you can do `window.location = '?foo=bar'`, which is the same as `/?foo=bar` (it refreshes the page)

you can also have `<a href="?foo=bar"></a>`

File: tests/NavigationTests.elm (hunk 16 above)
Comment (rogeriochaves) [reply tests/NavigationTests.elm]: It stays at the same place, that was what I was actually trying to test here, going back a hundred times to see how it behaves at the beginning of history, but perhaps would be better if we added a back/forward function like you said above to exercise the whole flow instead of calling the Msgs directly

File: tests/NavigationTests.elm (hunk 17 above)
Comment (rogeriochaves) [reply tests/NavigationTests.elm]: Yes, here I've tried to do more higher level tests here, testing things integrated, with `testable/NavigationTests` on a more unitary level, also to make it easier to extract the location encoding/parsing to a lib later

File: tests/Testable/NavigationTests.elm (hunk 18 above)
Comment (rogeriochaves) [reply tests/Testable/NavigationTests.elm:250]: Yes, like I mentioned above, those are more unitary and the others more integration, I believe both are important, here we catch the edge cases (and there are a lot of them when handling urls), while there we check that everything is working well together (also there we test the helpers functions that only live on TestContext, like `|> navigate "/foo"`)

File: tests/NavigationTests.elm (hunk 16 above)
Comment (avh4) [reply tests/NavigationTests.elm]: 👍   though I think we do also still want a couple tests like this to check that navigation Cmds generated by the program itself do the right thing.

File: tests/NavigationTests.elm (hunk 12 above)
Comment (rogeriochaves) [reply tests/NavigationTests.elm]: actually, this wasn't even being used on the tests anymore, I'll just remove it

File: examples/NavigationSample.elm (hunk 1 above)
Comment (rogeriochaves) [reply examples/NavigationSample.elm]: this file is just a copy of https://github.com/elm-lang/navigation/blob/master/examples/Example.elm, even this comment there is not mine :P

I'll replace with this other (more complete) one:
https://github.com/evancz/url-parser/blob/master/examples/Example.elm

File: src/Native/Testable/Navigation.js (hunk 5 above)
Comment (rogeriochaves) [reply src/Native/Testable/Navigation.js:39]: It doesn't, I don't remember exactly why right now, maybe loading order or idk, I was using this at first but had to overwrite both anyway so I got rid of Native_Navigation in [this refactor](https://github.com/rogeriochaves/elm-testable/commit/7b1f64aa31662308c23168f7a28e94189046cd97). Maybe we should try it again later.

File: src/Native/Testable/Navigation.js (hunk 8 above)
Comment (rogeriochaves) [reply src/Native/Testable/Navigation.js:56]: I've used the same names from https://github.com/elm-lang/navigation/blob/master/src/Navigation.elm#L240 so the codes could be closer. Should I still change?

File: src/Testable/Navigation.elm (hunk 6 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm]: yeah, so, one problem is that both `Html.program` and `Navigation.program` return a `Program Never Model Msg`, I don't know a way to tell them apart.

also, I believe you could use things like `Navigation.back` on a regular `Html.program`, the difference is that by using `Navigation.program` you can listen to browser url change events.

the thing is: an application always has a url, even a non-navigational one, so what should we do? maybe add a `startWithLocation` function but fallback to the default url if one not provided?

Comment (avh4) [reply src/Testable/Navigation.elm]: hmm.. If you use `Navigation.back` with a regular program, won't it always leave the program (or is there some way to intercept browser navigation without using `Navigation.program`?)

Even though both `Html.program` and `Navigation.program` return the same Elm types, we can still store hidden information in the javascript object so we can tell them apart at runtime.  This is how things currently work with flags (even though in the case of flags, the compiler knows something at the type level, there is a runtime check and some error handling logic that is performed at runtime).

File: src/Native/TestContext.js (hunk 4 above)
Comment (rogeriochaves) [reply src/Native/TestContext.js]: Hmmm I like the fallback because it is less native code from other parts to intercept (in this case, Html.program)

And what if there are other programs we don't know of? idk

File: tests/NavigationTests.elm (hunk 13 above)
Comment (rogeriochaves) [reply tests/NavigationTests.elm]: done

File: tests/NavigationTests.elm (hunk 14 above)
Comment (rogeriochaves) [reply tests/NavigationTests.elm]: Done, I believe taking a string is easier for the user to use

File: src/Testable/Navigation.elm (hunk 6 above)
Comment (rogeriochaves) [reply src/Testable/Navigation.elm]: Ok, I've added `startWithLocation` and `startWithFlagsAndLocation`. This increased complexity a bit, because now we need to save the original `init` function, that's because when the user calls `Navigation.Program` (which is before using `TestContext.start`) the init function is called with the current url.

So I store the original init function and call it again with the url from `startWithLocation`.

Still, it is possible for the user to use the `Navigation.Program` with the `TestContext.start` function (without the withLocation) because like I said, the types signatures are the same. In this case, the url fallbacks to `https://elm.testable/`

What do you think?

**Comment by rogeriochaves** [comment]
Do you believe we have enough for merging this?

**Comment by rogeriochaves** [comment]
@avh4 ping
